src/tag.rs

```diff
@@ -58,183 +81,231 @@ impl Tag {
+    ///     have value 1.
+    ///     Second and following octets give the tag number, base 128, most significant digit first,
+    ///     with as few digits as possible, and with the bit 8 of each octet except the last set
+    ///     to 1.
+    //
+    /// # Panics
+    ///
+    /// Currently, this function panics if the tag number is greater than
+    /// `MAX_VAL_SPAN_3_OCTETS`.
+    #[inline]
+    fn new(number: u32, type_mask: u8) -> Self {
+        assert!(number <= Tag::MAX_VAL_SPAN_3_OCTETS);
+        if number <= Tag::MAX_VAL_FOURTH_OCTET {
+            Tag([type_mask | number as u8, 0, 0, 0])
+        } else if number <= Tag::MAX_VAL_SPAN_1_OCTET {
+            // Fit the number in the third octets
+            let number = number as u8;
+            Tag([type_mask | Tag::SINGLEBYTE_DATA_MASK, number, 0, 0])
+        } else if number <= Tag::MAX_VAL_SPAN_2_OCTETS {
+            // Fit the number in the second and the third octets
+            let first_part = Tag::MULTIBYTE_DATA_MASK & ((number >> 7) as u8) | Tag::LAST_OCTET_MASK;
+            let second_part = Tag::MULTIBYTE_DATA_MASK & (number as u8);
+            Tag([type_mask | Tag::SINGLEBYTE_DATA_MASK, first_part, second_part, 0])
+        } else {
+            // Fit the number in the first, second and the third octets
+            let first_part = Tag::MULTIBYTE_DATA_MASK & ((number >> 14) as u8) | Tag::LAST_OCTET_MASK;
+            let second_part = Tag::MULTIBYTE_DATA_MASK & ((number >> 7) as u8) | Tag::LAST_OCTET_MASK;
+            let third_part = Tag::MULTIBYTE_DATA_MASK & (number as u8);
+            Tag([type_mask | Tag::SINGLEBYTE_DATA_MASK, first_part, second_part, third_part])
+        }
+    }
     /// Creates a new tag in the universal class with the given tag number.
     ///
     /// # Panics
     ///
     /// Currently, this function panics if the tag number is greater than
-    /// 30.
-    pub fn universal(number: u8) -> Self {
-        assert!(number < 31);
-        Tag(number)
+    /// `MAX_VAL_SPAN_3_OCTETS`.
+    pub fn universal(number: u32) -> Self {
+        Tag::new(number, Tag::UNIVERSAL)
     }
 
     /// Creates a new tag in the application class with the given tag number.
     ///
     /// # Panics
     ///
     /// Currently, this function panics if the tag number is greater than
-    /// 30.
-    pub fn application(number: u8) -> Self {
-        assert!(number < 31);
-        Tag(Tag::APPLICATION | number)
+    /// `MAX_VAL_SPAN_3_OCTETS`.
+    pub fn application(number: u32) -> Self {
+        Tag::new(number, Tag::APPLICATION)
     }
 
     /// Creates a new tag in the context specific class.
     ///
     /// # Panics
     ///
     /// Currently, this function panics if the provided tag number is greater
-    /// than 30.
-    pub fn context_specific(number: u8) -> Self {
-        assert!(number < 31);
-        Tag(Tag::CONTEXT_SPECIFIC| number)
+    /// than `MAX_VAL_SPAN_3_OCTETS`.
+    pub fn context_specific(number: u32) -> Self {
```

Comment:
One more thing: If you rename this to `ctx` (or maybe `context`, if `ctx` is too cryptic?), this could be used instead of the consts, i.e., `Tag::ctx(2)` instead of `Tag::CTX_2` which would make it consistent beyond the defined consts.

Reply:
Changed.